### PR TITLE
Implement signing scheme.

### PR DESCRIPTION
**Pull request by priteshbandi**
### Change Log
1. Implement signature scheme. Specs in https://github.com/notaryproject/notaryproject/pull/165
2. address comment of PR #15

### Testing
Unit tests.

### Signature example
Minimal Signature: https://tinyurl.com/878aj5vd
Loaded Signature: https://tinyurl.com/2p8397vb

### CODE REVIEW

File: signer/signer.go
Comment (rgnote) [inline signer/signer.go:50]:
```suggestion
	VerificationPluginMinVersion string  // TODO: Implement SemVer structure
```

File: signer/jws.go
Comment (rgnote) [inline signer/jws.go:204]: nit: No need to use a pointer reference here and below

File: signer/jws.go
Comment (rgnote) [inline signer/jws.go:266]:
```suggestion
		return MalformedSignatureError{fmt.Sprintf("these required headers are not marked as critical : %v", keys)}
```

File: signer/jws.go
Comment (rgnote) [inline signer/jws.go:360]:
```suggestion
	// The time at which the signature was generated. only valid when signing scheme is `notary.default.x509`
```

File: signer/jws.go
Comment (rgnote) [inline signer/jws.go:363]:
```suggestion
	// The time at which the signature was generated. only valid when signing scheme is `notary.signingAuthority.x509`
```

File: signer/jws.go
Comment (rgnote) [inline signer/jws.go:366]:
```suggestion
	// VerificationPlugin specifies the name of the verification plugin that should be used to verify the signature.
```

File: signer/jws.go
Comment (rgnote) [inline signer/jws.go:369]:
```suggestion
	// VerificationPluginMinVersion specifies the minimum version of the verification plugin that should be used to verify the signature.
```

File: signer/jws.go
Comment (rgnote) [inline signer/jws.go:370]:
```suggestion
	// TODO: replace string with well defined struct for SemVer.
```

File: signer/signer.go
Comment (rgnote) [reply signer/signer.go:50]: We can accept a string here rather than a struct. The TODO part is validating that the SemVer string is valid

File: signer/signer.go
Comment (rgnote) [inline signer/signer.go:155]: Just pass the info struct rather than individual fields?

File: signer/jws.go
Comment (priteshbandi) [reply signer/jws.go:204]: Here we are copying data refereed by `protectedHdr.Expiry` pointer to `signInfo.SignedAttributes.Expiry`. 
`protectedHdr.Expiry` needs to be a pointer because expiry is optional field and shouldn't be present in signature if user haven't specified expiry.

File: signer/signer.go
Comment (priteshbandi) [reply signer/signer.go:155]: We are reusing validate function across signing and verification call. Both have different structure(SignRequest and SignerInfo)

File: signer/signer.go
Comment (priteshbandi) [reply signer/signer.go:50]: As we need to validate semver which has its own semantics and compare with other semver, IMO it would be better to have well defined struct and fail early.

Comment (priteshbandi) [reply signer/signer.go]: Since golang doesnt provides require library ([semver in x pkg](https://pkg.go.dev/golang.org/x/mod@v0.5.1/semver)) doest abide by https://semver.org/. Used regular expression to validate semver.

**Review by rgnote — APPROVED** [review]

File: signer/types.go
Comment (rgnote) [inline signer/types.go:72]: These were changed to `notary.x509` and `notary.x509.signingAuthority` in the latest spec. See https://github.com/notaryproject/notaryproject/blob/main/signature-specification.md#standard-attributes

Comment (priteshbandi) [reply signer/types.go:72]: Yep, waiting for @gokarnm to review and if there are any feedback will squash this renaming with other changes.

**Comment by dtzar** [comment]
Per the [PR review criteria](https://github.com/notaryproject/notation/discussions/263) @notaryproject/notation-sub-group-maintainers - this is past a week old, BUT needs one more approval from @notaryproject/notation-code-reviewers, to be able to be merged.

**Review by shizhMSFT — APPROVED** [review]
LGTM

File: signer/jws.go
Comment (gokarnm) [inline signer/jws.go:259]: Nit, shouldn't this comment be in line 248? Here we throw an error if header is present but not marked critical.

File: signer/jws.go
Comment (gokarnm) [inline signer/jws.go:237]: For Signing scheme X509, presence of `AuthenticSigningTime` is invalid. And for X509 Signing Authority, presence of `SigningTime` is invalid. We should add those validations.

https://github.com/notaryproject/notaryproject/blob/main/signature-specification.md#signed-attributes

File: signer/signer.go
Comment (gokarnm) [inline signer/signer.go:22]: Nit, I'd be preferable to organize these as signed and unsigned and have some comments.

```
// SignerInfo represents a parsed signature envelope that is agnostic to signature envelope format.
type SignerInfo struct {
     
       Payload            []byte

        // Signed attributes
        SignatureAlgorithm SignatureAlgorithm
	SigningScheme      SigningScheme
	PayloadContentType PayloadContentType
	SignedAttributes   SignedAttributes

       // The signature is generated over payload and signed attributes
       Signature          []byte

       // Unsigned attributes
       CertificateChain   []*x509.Certificate
	TimestampSignature []byte
	UnsignedAttributes UnsignedAttributes
	SignatureAlgorithm SignatureAlgorithm
}
```

File: signer/signer.go
Comment (gokarnm) [inline signer/signer.go:211]: Nit, I'd use full names for vPlugin, vPluginVersion

File: signer/jws.go
Comment (priteshbandi) [reply signer/jws.go:259]: Nop, this is the correct comment. We are doing two validations
1. Validate that all headers market as critical are present in the headers map.
2. Validate that all required headers(as per spec) are marked critical.

**Review by gokarnm — APPROVED** [review]
LGTM!

**Pull request merged**
